File: sources/CPP/include/Ers/SubModel/SubModelRandomProperties.cpp
#include "SubModelRandomProperties.h"

#include "Ers/Api.h"

namespace Ers
{
    void SubModelRandomProperties::SetRandomGenerator(Ers::RandomGeneratorType randomNumberGeneratorType)
    {
        Ers::Engine::ERS_SubModelRandomProperties_SetRandomGenerator_Type(CorePtr(), static_cast<int>(randomNumberGeneratorType));
    }

    void SubModelRandomProperties::SetRepetitive(bool value)
    {
        Ers::Engine::ERS_SubModelRandomProperties_SetRepetitive(CorePtr(), value);
    }

    void SubModelRandomProperties::SetAntithetical(bool value)
    {
        Ers::Engine::ERS_SubModelRandomProperties_SetAntithetical(CorePtr(), value);
    }

    void SubModelRandomProperties::SetSeed(size_t seed)
    {
        Ers::Engine::ERS_SubModelRandomProperties_SetSeed(CorePtr(), seed);
    }

    bool SubModelRandomProperties::IsRepetitive()
    {
        return Ers::Engine::ERS_SubModelRandomProperties_IsRepetitive(CorePtr());
    }

    bool SubModelRandomProperties::IsAntithetical()
    {
        return Ers::Engine::ERS_SubModelRandomProperties_IsAntithetical(CorePtr());
    }

    [[nodiscard]] size_t SubModelRandomProperties::GetOriginalSeed() const
    {
        return Ers::Engine::ERS_SubModelRandomProperties_GetOriginalSeed(CorePtr());
    }

    [[nodiscard]] Ers::RandomGenerator& SubModelRandomProperties::GetRandomNumberGenerator()
    {
        void* coreRandomGeneratorInstance = Ers::Engine::ERS_SubModelRandomProperties_GetRandomNumberGenerator(CorePtr());
        return *static_cast<Ers::RandomGenerator*>(coreRandomGeneratorInstance);
    }

    void* SubModelRandomProperties::CorePtr()
    {
        return this;
    }

    const void* const SubModelRandomProperties::CorePtr() const
    {
        return this;
    }
} // namespace Ers

File: sources/CSharp/src/Ers/SubModel/SubModelRandomProperties.cs
using Ers.Engine;
using System.Runtime.InteropServices;

namespace Ers
{
    /// <summary>
    /// The properties of the <see cref="RandomGenerator"/> attached to a <see cref="SubModel"/>.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public ref struct SubModelRandomProperties
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public readonly IntPtr CorePtr;

        internal SubModelRandomProperties(IntPtr corePtr) { CorePtr = corePtr; }

        /// <summary>
        /// Set the random number generator by its type.
        /// </summary>
        /// <param name="randomNumberGeneratorType">The type of the new random number generator. See <see
        /// cref="RandomGeneratorType"/>.</param>
        public void SetRandomGenerator(RandomGeneratorType randomNumberGeneratorType)
        {
            ErsEngine.ERS_SubModelRandomProperties_SetRandomGenerator_Type(CorePtr, (int)randomNumberGeneratorType);
        }

        /// <summary>
        /// Get the random number generator attached to the <see cref="SubModel"/>.
        /// </summary>
        /// <returns></returns>
        public RandomGenerator GetRandomNumberGenerator()
        {
            IntPtr coreRandomGeneratorInstance = ErsEngine.ERS_SubModelRandomProperties_GetRandomNumberGenerator(CorePtr);

            unsafe
            {
                return new RandomGenerator(coreRandomGeneratorInstance);
            }
        }
    }
}
